roles/expirationd.lua

```diff
@@ -0,0 +1,302 @@
+        for _, func in pairs(missed_functions_opts) do
+            table.insert(missed_functions, func)
+        end
+    end
+    return conf, missed_functions
+end
+
+local function get_cfg(cfg)
+    local conf = setmetatable(table.deepcopy(cfg), {})
+    local params_map = {
+        metrics = {"b"},
+    }
+
+    for k, _ in pairs(conf) do
+        if type(k) ~= "string" then
+            error(role_name .. ": config option must be a string")
+        end
+        if params_map[k] == nil then
+            error(role_name .. ": unsupported config option " .. k)
+        end
+    end
+
+    for param, types in pairs(params_map) do
+        if conf[param] ~= nil then
+            local _, ok, err = get_param(param, conf[param], types)
+            if not ok then
+                error(err)
+            end
+        end
+    end
+
+    return conf
+end
+
+local function validate_config(conf_new)
+    local conf = conf_new or {}
+
+    for task_name, task_conf in pairs(conf) do
+        local _, ok, err = get_param("task name", task_name, {"s"})
+        if not ok then
+            error(err)
+        end
+        local _, ok, err = get_param("task params", task_conf, {"t"})
+        if not ok then
+            error(err)
+        end
+        local ok, ret = pcall(get_task_config, task_conf)
+        if not ok then
+            if task_name == "cfg" then
+                get_cfg(task_conf)
+            else
+                error(ret)
+            end
+        end
+    end
+
+    return true
+end
+
+local function load_task(task_conf, task_name)
+    local timeout = 1
+    fiber.name(role_name .. ": " .. task_name)
+    local start = fiber.clock()
+    local task_config, missed_functions = get_task_config(task_conf)
+    local skip = task_conf.is_master_only and not box.info.ro
+    if skip then
+        return
+    end
+    while #missed_functions ~= 0 do
+        fiber.sleep(timeout)
+        if fiber.clock() - start > 60 then
+            local message = role_name .. ": " .. task_name .. ": waiting for functions: "
+            for _, func in pairs(missed_functions) do
+                message = message .. func .. ', '
+            end
+            log.warn(message)
+            start = fiber.clock()
+        end
+        task_config, missed_functions = get_task_config(task_conf)
+    end
+
+    local task = expirationd.start(task_name, task_config.space,
+                                    task_config.is_expired,
+                                    task_config.options)
+    if task == nil then
+        error(role_name .. ": unable to start task " .. task_name)
+    end
+    table.insert(started, task_name)
+
+
+end
+
+local function apply_config(conf)
+    -- Finishes tasks from an old configuration
+    for i=#started,1,-1 do
+        local task_name = started[i]
+        local ok, _ = pcall(expirationd.task, task_name)
+        if ok then
+            if conf[task_name] then
+                expirationd.task(task_name):stop()
```

Comment:
I expect that in case of unchanged task config section the behavior shouldn't change. So the question is whether restart changes it or not.

Reply:
It's behavior is similar to cartridge role. If task is still in config it will restart, which means that it continues processing from a last tuple if the task has default start_key and iterate_with functions. I am not so sure about expiration timers, but i don't want to change behavior from cartridge.

Reply:
Timers reset on task restart, but i don't see it as a big problem, since it is behavior that was in cartridge. (and having a big timer is rare usecase) Also  we can not guarantee that if for example name of the function in config has not changed it is still the same. If you want you can write a wrapper to change that behavior.

Reply:
Ok, thank for the clear explanation. Sticking to existing approach is good.